diff --git a/panic_unwind/src/lib.rs b/panic_unwind/src/lib.rs
--- a/panic_unwind/src/lib.rs
+++ b/panic_unwind/src/lib.rs
@@ -30,6 +30,10 @@ use core::panic::PanicPayload;

 cfg_select! {
+    target_os = "jvm" => {
+        #[path = "jvm.rs"]
+        mod imp;
+    }
     target_os = "hermit" => {
         #[path = "hermit.rs"]
         mod imp;
     }
diff --git a/std/src/sys/personality/mod.rs b/std/src/sys/personality/mod.rs
--- a/std/src/sys/personality/mod.rs
+++ b/std/src/sys/personality/mod.rs
@@ -16,7 +16,7 @@ cfg_select! {
     target_os = "emscripten" => {
         mod emcc;
     }
-    any(target_env = "msvc", target_family = "wasm", target_os = "motor") => {
+    any(target_os = "jvm", target_env = "msvc", target_family = "wasm", target_os = "motor") => {
         // This is required by the compiler to exist (e.g., it's a lang item),
         // but it's never actually called by the compiler because
         // __CxxFrameHandler3 (msvc) / __gxx_wasm_personality_v0 (wasm) is the
